proposals/000-composite-fetcher-api.md

```diff
@@ -0,0 +1,357 @@
+
+#### Mobile (Swift & Kotlin)
+
+This proposal is a non-breaking change, since it describes a structure that is mostly internal. The public features are new, such as adding resource transformers and decorating resource access.
+
+
+## Reference Guide
+
+* [Examples of `Fetcher` Trees](#examples-of-fetcher-trees)
+* [`Fetcher` Interface](#fetcher-interface)
+* [`Resource` Interface](#resource-interface)
+* Leaf Fetchers
+  * [`FileFetcher`](#filefetcher-class)
+  * [`HTTPFetcher`](#httpfetcher-class)
+  * [`ZIPFetcher`](#zipfetcher-class)
+  * [`ProxyFetcher`](#proxyfetcher-class)
+* Composite Fetchers
+  * [`RoutingFetcher`](#routingfetcher-class)
+  * [`TransformingFetcher`](#transformingfetcher-class)
+  * [`CachingFetcher`](#cachingfetcher-class)
+
+### Examples of `Fetcher` Trees
+
+The fetcher tree created by the publication parsers can be adapted to fit the characteristics of each format.
+
+#### CBZ and ZAB (Zipped Audio Book)
+
+These formats are very simple, we just need to access the ZIP entries.
+
+<img src="assets/002-zip.svg">
+
+#### Audiobook Manifest
+
+The resources of a remote audiobook are fetched with HTTP requests, using an `HTTPFetcher`. However, we can implement an offline cache by wrapping the fetcher in a `CachingFetcher`.
+
+<img src="assets/002-audiobook.svg">
+
+#### LCP Protected Package (Audiobook, LCPDF, etc.)
+
+The resources of a publication protected with LCP need to be decrypted. For that, we're using a `DecryptionTransformer` embedded in a `TransformingFetcher`. Any remote resources declared in the manifest are fetched using an `HTTPFetcher`.
+
+<img src="assets/002-lcp.svg">
+
+#### EPUB
+
+The EPUB fetcher is one of the most complex:
+* An `HTTPFetcher` is used for [remote resources](https://www.w3.org/publishing/epub3/epub-spec.html#sec-resource-locations).
+* The resources are transformed at two different levels:
+  * in the *streamer*, to decrypt the content and deobfuscate fonts,
+  * in the *navigator*, to inject the CSS and JavaScript necessary for rendering.
+
+<img src="assets/002-epub.svg">
+
+### `Fetcher` Interface
+
+Provides access to a `Resource` from a `Link`.
+
+#### Methods
+
+* `get(link: Link, parameters: Map<String, String> = {}) -> Resource`
+  * Returns the `Resource` at the given `Link.href`.
+    * A `Resource` is *always* returned, since for some cases we can't know if it exists before actually fetching it, such as HTTP. Therefore, errors are handled at the `Resource` level.
+  * `link: Link`
+    * Link targeting the resource.
+    * We're expecting a `Link` because a `Fetcher` might use its properties, e.g. to transform the resource. Therefore, `Publication::get()` makes sure that a complete `Link` is always provided to the `Fetcher`.
+  * `parameters: Map<String, String> = {}`
+    * HREF parameters that the source will understand, such as:
+      * when `Link::href` is templated,
+      * to append additional query parameters to an HTTP request.
```

Comment:
How could this mix of usages be implemented?
Assume href is `/search?q={searchTerms}`. You will replace `searchTerms` by `parameters["searchTerms"]`, that's ok. Would any other parameter be appended as "&parameter.key=parameter.value"? What in case of conflict, i.e. /search?q={searchTerms}&searchTerms=value?

Reply:
I guess the URI template should take precedence, so we can pop each parameters to be expanded, and then append the rest to the URI.

Reply:
I didn't write about it in the proposal, but these parameters can be "2-ways", meaning that we can receive some query parameters from an HTTP request to a Publication Service's WS API, which will be passed in this `parameters` argument to be used by the Publication Service implementation. But this can also be used, as described, to send query parameters through an `HTTPFetcher`, either by expanding a URI template, or by adding more query parameters.

So `parameters` may have been parsed from query params, and may be serialized to query params. I say *may* because a Publication Service will use the `parameters` map directly.